packages/neon-core/src/tx/transaction/Transaction.ts

```diff
@@ -1,104 +1,386 @@
-    ss: StringStream,
-    tx: Partial<TransactionLike>
-  ): Partial<TransactionLike> {
-    throw new Error("Method not implemented.");
+  /**
+   * Add an attribute.
+   * @param usage The usage type. Do refer to txAttrUsage enum values for all available options.
+   * @param data The data as hexstring.
+   */
+  public addAttribute(usage: number, data: string): this {
+    if (typeof data !== "string") {
+      throw new TypeError("data should be formatted as string!");
+    }
+    this.attributes.push(new TransactionAttribute({ usage, data }));
+    return this;
   }
 
-  private constructor(tx: Partial<TransactionLike> = {}) {
-    super(tx);
+  /**
+   * Adds an Witness to the Transaction and automatically sorts the witnesses according to scripthash.
+   * @param witness The Witness object to add.
+   */
+  public addWitness(witness: Witness): this {
+    if (witness.scriptHash === "") {
+      throw new Error("Please define the scriptHash for this Witness!");
+    }
+    this.scripts.push(witness);
+    this.scripts = this.scripts.sort(
+      (w1, w2) => parseInt(w1.scriptHash, 16) - parseInt(w2.scriptHash, 16)
+    );
+    return this;
   }
 
-  public get [Symbol.toStringTag](): string {
-    return "Transaction";
+  /**
+   * Adds some script intents to the Transaction
+   * System Fee will be increased automatically
+   * However system Fee calculated in this method is insufficient if calling non-native contracts.
+   * If the transaction is invoking contracts other than native contracts, invokeScript rpc request can test the systemFee.
+   * @param scriptIntents sciprt Intents to add to the transaction
+   */
+  public addIntent(...scriptIntents: ScriptIntent[]): this {
+    let increasedSystemFee = new Fixed8(0);
+    this.script = scriptIntents.reduce((accumulatedScript, intent) => {
+      const { hex, fee } = createScript(intent);
+      this._pre_systemFee = this._pre_systemFee.plus(fee);
+      increasedSystemFee = increasedSystemFee.plus(fee);
+      this.intents.push(intent);
+      return accumulatedScript + hex;
+    }, this.script);
+    log.info(
+      `Increased systemFee: ${increasedSystemFee.toNumber()}, totally ${this.systemFee.toNumber()}`
+    );
+    return this;
   }
 
   /**
-   * Exclusive Data
+   * If transaction script is constructed only by method `addIntent`, systemFee will be calcuated.
+   * If all intents are about native contract invocation, this system fee is accurate.
+   * In this case, you can use this method to get minimum system fee.
    */
-  public get exclusiveData(): { [key: string]: any } {
-    throw new Error("Not Implemented!");
+  public useCalculatedSystemFee(): void {
+    this.systemFee = this._pre_systemFee;
   }
 
-  public get fees(): number {
-    return 0;
+  /**
+   * This function regulate system fee with 2 processes:
+   * 1. considering the free system fee threshold
+   * 2. ceil systemFee with factor
+   */
+  public regulateSystemFee(): void {
```

Comment:
please refactor + explain the logic more.

Reply:
explanation updated.

Reply:
is there not a ceiling method?